### PR TITLE
Et3 5 et3 8

### PR DESCRIPTION
**Pull request by SkoriyVladislav**
ET3-5, ET3-8

### CODE REVIEW

**Comment by SkoriyVladislav** [comment]
ET3-5, ET3-8

File: back-end/src/main/resources/email_sender/mail.properties [hunk 1]
```diff
@@ -0,0 +1,7 @@
+mail.smtp.auth=true
+mail.smtp.starttls.enable=true
+mail.smtp.host=smtp.gmail.com
+mail.smtp.port=587
```
Comment (SergeyKoval) [inline back-end/src/main/resources/email_sender/mail.properties]: move properties to the spring boot application.properties

File: back-end/src/main/java/com/exadel/team3/backend/service/email/email_sender/impl/SenderImpl.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+@Component
+public class SenderImpl implements Sender {
+    private String username;
+    private String password;
+
+    public SenderImpl() {
+        try {
+            Properties properties = new Properties();
+            InputStream inputStream = this.getClass().getClassLoader().getResourceAsStream("email_sender/mail.properties");
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/service/email/email_sender/impl/SenderImpl.java]: Use @Value to get property

File: back-end/src/main/java/com/exadel/team3/backend/service/email/email_sender/impl/SenderImpl.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    public boolean send(EmailTypes emailType, String toEmail) {
+
+        try {
+            final Properties properties = new Properties();
+            properties.load(Sender.class.getClassLoader().getResourceAsStream("email_sender/mail.properties"));
+
+            Session session = Session.getInstance(properties, new Authenticator() {
+                protected PasswordAuthentication getPasswordAuthentication() {
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/service/email/email_sender/impl/SenderImpl.java]: No anonymous classes!!! Why not lambda?

File: back-end/src/main/java/com/exadel/team3/backend/service/email/email_sender/impl/SenderImpl.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+            Message message = new MimeMessage(session);
+            message.setFrom(new InternetAddress(username));
+            message.setRecipients(Message.RecipientType.TO, InternetAddress.parse(toEmail));
+            message.setSubject(EmailCommandFactory.getEmailSubject(emailType, toEmail));
+            message.setContent(EmailCommandFactory.getEmailText(emailType, toEmail), "text/html; charset=utf-8");
+            Transport.send(message);
+            return true;
+        } catch (IOException | MessagingException ex) {
+            return false;
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/service/email/email_sender/impl/SenderImpl.java]: At least log an exception

File: back-end/src/main/java/com/exadel/team3/backend/service/email/email_command/email_command_type/EmailTypes.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.exadel.team3.backend.service.email.email_command.email_command_type;
+
+public enum EmailTypes {
+    CONFIRM_EMAIL,
+    SEND_LOGIN_AND_PASS,
+    VERIFICATION_TEST;
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/service/email/email_command/email_command_type/EmailTypes.java]: Try to put implementation right there (we can discuss it separately)

File: back-end/src/main/java/com/exadel/team3/backend/service/email/email_command/email_command_impl/VerificationTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        return text;
+    }
+
+    @Override
+    public String getEmailText(String email) throws IOException{
+
+        // TODO get link form BD
+        String link = "link";
+        final Properties properties = new Properties();
+        properties.load(VerificationTest.class.getClassLoader().getResourceAsStream("email_sender/text.properties"));
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/service/email/email_command/email_command_impl/VerificationTest.java]: a lot of copy paste in each impleemntation

File: back-end/src/main/java/com/exadel/team3/backend/service/email/email_command/email_command_impl/ConfirmEmail.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        try (BufferedReader reader = new BufferedReader(inputStreamReader)){
+            String line;
+            while ((line = reader.readLine()) != null) {
+                text += line;
+            }
+        } catch (IOException e) {
+            //TODO Exception here
+        }
+
+        text = text + "<a href=\"" + email + "?id=" + checker + "\" >" + email + "?id=" + checker + "</a>";
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/service/email/email_command/email_command_impl/ConfirmEmail.java]: Don't use such concatenations!

File: back-end/src/main/java/com/exadel/team3/backend/services/email/EmailConfig.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import com.exadel.team3.backend.services.email.email_sender.Sender;
+import com.exadel.team3.backend.services.email.email_sender.impl.SenderImpl;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.*;
+
+@Configuration
+@ComponentScan("com.exadel.team3.backend.services.email")
+public class EmailConfig {
+    @Autowired
+    private Sender sender;
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/EmailConfig.java]: Why do you need this class?

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/Sender.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.exadel.team3.backend.services.email.email_sender;
+
+
+import com.exadel.team3.backend.services.email.email_types.EmailTypes;
+
+import java.util.Map;
+
+public interface Sender {
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/Sender.java]: Rename to mail sender

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+import javax.mail.*;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Properties;
+
+@Component
+@PropertySource("classpath:application.properties")
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java]: You are able to get application properties without this

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    private String password;
+
+    public SenderImpl() {
+    }
+
+    public boolean send(EmailTypes emailType, String toEmail, Map<String, String> replacementMap) {
+
+        try {
+            final Properties properties = new Properties();
+            properties.load(Sender.class.getClassLoader().getResourceAsStream("application.properties"));
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java]: Why again manually loading of application.properties

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+            Message message = new MimeMessage(session);
+            message.setFrom(new InternetAddress(username));
+            message.setRecipients(Message.RecipientType.TO, InternetAddress.parse(toEmail));
+            message.setSubject(emailType.getTemplateSubject());
+            message.setContent(replaseVariabl(emailType.getTemplateText(),replacementMap), "text/html; charset=utf-8");
+            Transport.send(message);
+            return true;
+        } catch (IOException | MessagingException ex) {
+            return false;
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java]: Log exception

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    private String replaseVariabl(String text, Map<String, String> replacementMap) {
+        for(Map.Entry<String, String> entry : replacementMap.entrySet()) {
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_sender/impl/SenderImpl.java]: Use Java 8 style

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_types/EmailTypes.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
+package com.exadel.team3.backend.services.email.email_types;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+public enum EmailTypes {
+
+    CONFIRM_EMAIL("email_sender/text_for_email/Text_for_confirm_email.txt", "email_sender/subject_for_email/Subject_for_congirm_email.txt"),
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_types/EmailTypes.java]: Move this part to the constant as long as it is the same for all types:
email_sender/text_for_email/

File: back-end/src/main/java/com/exadel/team3/backend/services/email/email_types/EmailTypes.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    private String readFile(String nameFile) {
+        InputStreamReader inputStreamReader = new InputStreamReader(this.getClass().getClassLoader().getResourceAsStream(nameFile));
+        String text = "";
+        try (BufferedReader reader = new BufferedReader(inputStreamReader)){
+            String line;
+            while ((line = reader.readLine()) != null) {
+                text += line;
+            }
+        } catch (IOException e) {
+            //TODO Exception here
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/email/email_types/EmailTypes.java]: Log an exception

File: back-end/src/main/resources/email_sender/subject_for_email/Subject_for_congirm_email.txt [hunk 16]
```diff
@@ -0,0 +1 @@
+Подтверждение почтового ящика для сайта CATS
```
Comment (SergeyKoval) [inline back-end/src/main/resources/email_sender/subject_for_email/Subject_for_congirm_email.txt]: Use English text in all messages. We are international company =)

File: back-end/src/main/java/com/exadel/team3/backend/services/mail/mail_sender/MailAuthenticator.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
+package com.exadel.team3.backend.services.mail.mail_sender;
+
+import javax.mail.Authenticator;
+import javax.mail.PasswordAuthentication;
+
+public class MailAuthenticator extends Authenticator {
+    private String username;
+    private String password;
+    private static MailAuthenticator mailAuthenticator;
```
Comment (SergeyKoval) [inline back-end/src/main/java/com/exadel/team3/backend/services/mail/mail_sender/MailAuthenticator.java]: no need to for static variable - it needs to be not a singleton. Just public static init method, which will call private constructor and directly return class instance

**Pull request merged**